<a href="https://colab.research.google.com/github/ashennavindu/Statistical-Learning-e23022/blob/main/Assignment%202/Q3%2CQ4%2CQ4%2CQ5%2CQ6%2CQ7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Q3) Transformer Failiures

In a certain electricity distribution district, the number of emergency transformer failures reported in a single day is modeled as a discrete random variable $X$. Assume that $X$ takes values in the set $\{0,1,2,3\}$, with probabilities

$$
P(X=0)=0.50,\qquad P(X=1)=0.30,\qquad P(X=2)=0.15,\qquad P(X=3)=0.05.
$$

1. Construct a probability space $(\Omega,\mathfrak F,P)$ suitable for this model.
2. Define $X$ as a random variable on this space.
3. Determine the probability mass function of $X$.
4. Compute the expectation $E[X]$.
5. Compute the variance $\mathrm{Var}(X)$.
6. Find $P(X\ge 1)$ and $P(X\ge 2)$.
7. Briefly explain the practical meaning of the expectation and variance in this setting.



### Solutions for Q3: Transformer Failures

**1. Probability Space $(\Omega, \mathfrak{F}, P)$:**
* **Sample Space ($\Omega$):** $\Omega = \{0, 1, 2, 3\}$, representing the possible number of failures.
* **Sigma-algebra ($\mathfrak{F}$):** The power set of $\Omega$, i.e., $\mathfrak{F} = \mathcal{P}(\Omega)$, which contains all $2^4 = 16$ possible events.
* **Probability Measure ($P$):** Defined by the given weights: $P(\{0\})=0.50, P(\{1\})=0.30, P(\{2\})=0.15, P(\{3\})=0.05$.

**2. Random Variable $X$:**
$X: \Omega \to \mathbb{R}$ is the identity mapping $X(\omega) = \omega$ for all $\omega \in \Omega$.

**3. Probability Mass Function (PMF):**
$$p_X(x) = \begin{cases} 0.50 & x=0 \\ 0.30 & x=1 \\ 0.15 & x=2 \\ 0.05 & x=3 \\ 0 & \text{otherwise} \end{cases}$$

In [1]:
import pandas as pd
import numpy as np

# Data definition
x_values = np.array([0, 1, 2, 3])
probs = np.array([0.50, 0.30, 0.15, 0.05])

# 4. Compute Expectation E[X]
expectation = np.sum(x_values * probs)

# 5. Compute Variance Var(X)
# Var(X) = E[X^2] - (E[X])^2
expectation_x2 = np.sum((x_values**2) * probs)
variance = expectation_x2 - (expectation**2)

# 6. Probabilities P(X >= 1) and P(X >= 2)
p_ge_1 = np.sum(probs[x_values >= 1])
p_ge_2 = np.sum(probs[x_values >= 2])

print(f"4. Expectation E[X]: {expectation:.2f}")
print(f"5. Variance Var(X): {variance:.4f}")
print(f"6. P(X >= 1): {p_ge_1:.2f}")
print(f"   P(X >= 2): {p_ge_2:.2f}")

4. Expectation E[X]: 0.75
5. Variance Var(X): 0.7875
6. P(X >= 1): 0.50
   P(X >= 2): 0.20


**7. Practical Meaning:**
* **Expectation ($E[X]$):** Represents the long-term average number of emergency failures reported per day. If we observed this district over many years, the average daily failures would converge to this value.
* **Variance ($\text{Var}(X)$):** Measures the spread or uncertainty of the failures. A higher variance suggests that the number of failures fluctuates significantly from the average, making daily resource planning more difficult.

# Q4) Probability of Defective Items

In a manufacturing process, the proportion of defective items in a large batch produced on a given day is modeled by a continuous random variable $X$. Since $X$ represents a proportion, it takes values in the interval $[0,1]$. Suppose that $X$ has a Beta distribution with parameters $\alpha=2$ and $\beta=18$, so that its density is

$$
f(x)=
\begin{cases}
\dfrac{1}{B(2,18)}x^{1}(1-x)^{17}, & 0<x<1,\\\
0, & \text{otherwise},
\end{cases}
$$

where $B(2,18)$ is the Beta function.

Answer the following:

1. Construct a probability space $(\Omega,\mathfrak F,P)$ suitable for this model.

2. Define $X$ as a continuous random variable on this space.

3. Write down the probability density function of $X$.

4. Verify that $f$ is a valid probability density function.

5. Compute the expectation $E[X]$.

6. Compute the variance $\mathrm{Var}(X)$.

7. Find the probability that the defective proportion is at most $0.10$, that is, compute

$$
P(X\le 0.10).
$$

You may leave your answer in Beta integral form.

8. Find the probability that the defective proportion lies between $0.05$ and $0.15$, that is, compute

$$
P(0.05\le X\le 0.15).
$$

You may leave your answer in integral form.

9. Briefly explain why a Beta distribution is a reasonable model for this situation.

10. Interpret the expectation and variance in the context of production quality.


### Solutions for Q4: Probability of Defective Items

**1. Probability Space $(\Omega, \mathfrak{F}, P)$:**
* **Sample Space ($\Omega$):** $\Omega = [0, 1]$, representing all possible proportions of defective items.
* **Sigma-algebra ($\mathfrak{F}$):** The Borel sigma-algebra on $[0, 1]$, denoted $\mathcal{B}([0, 1])$.
* **Probability Measure ($P$):** Defined by the integral of the Beta density function over sets in $\mathfrak{F}$.

**2. Random Variable $X$:**
$X$ is a continuous random variable defined as the identity map $X(\omega) = \omega$ for $\omega \in [0, 1]$, which represents the proportion of defective items.

**3. Probability Density Function (PDF):**
Given $\alpha=2, \beta=18$:
$$f(x) = \frac{x^{2-1}(1-x)^{18-1}}{B(2, 18)} = \frac{x(1-x)^{17}}{B(2, 18)}, \quad 0 < x < 1$$

**4. Verification:**
$f(x) \ge 0$ for all $x \in (0, 1)$ and $\int_0^1 f(x)dx = 1$ by the definition of the Beta function $B(\alpha, \beta) = \int_0^1 t^{\alpha-1}(1-t)^{\beta-1}dt$.

In [2]:
from scipy.stats import beta

alpha_param = 2
beta_param = 18

# 5. Expectation E[X] = alpha / (alpha + beta)
expectation_q4 = beta.mean(alpha_param, beta_param)

# 6. Variance Var(X) = (alpha * beta) / ((alpha + beta)^2 * (alpha + beta + 1))
variance_q4 = beta.var(alpha_param, beta_param)

# 7. P(X <= 0.10)
p_le_010 = beta.cdf(0.10, alpha_param, beta_param)

# 8. P(0.05 <= X <= 0.15)
p_range = beta.cdf(0.15, alpha_param, beta_param) - beta.cdf(0.05, alpha_param, beta_param)

print(f"5. Expectation E[X]: {expectation_q4:.4f}")
print(f"6. Variance Var(X): {variance_q4:.6f}")
print(f"7. P(X <= 0.10): {p_le_010:.4f}")
print(f"8. P(0.05 <= X <= 0.15): {p_range:.4f}")

5. Expectation E[X]: 0.1000
6. Variance Var(X): 0.004286
7. P(X <= 0.10): 0.5797
8. P(0.05 <= X <= 0.15): 0.5562


**9. Reasoning for Beta Distribution:**
The Beta distribution is ideal for modeling proportions or percentages because its support is strictly bounded between 0 and 1. By adjusting $\alpha$ and $\beta$, it can take various shapes (skewed, symmetric, or U-shaped) to represent different manufacturing quality scenarios.

**10. Interpretation:**
* **Expectation ($E[X] = 0.10$):** On average, 10% of items in a batch are expected to be defective.
* **Variance:** Indicates the consistency of the production process. A low variance means the defective rate is stable across different days, while high variance suggests poor quality control and unpredictable defect rates.

# Q5) Expected Project Completion Time

A construction company is estimating the time required to install a backup generator at a hospital. Let $X$ denote the number of hours needed to complete the installation. Based on expert judgment, the shortest possible time is $4$ hours, the longest possible time is $10$ hours, and the most likely completion time is $6$ hours. Assume that $X$ follows a Triangular distribution with parameters $a=4$, $b=10$, and $c=6$.

Answer the following:

1. Specify the support of $X$.

2. Write down the probability density function $f_X(x)$.

3. Verify that $f_X$ integrates to $1$ over its support.

4. Compute the expectation $E[X]$.

5. Compute the variance $\mathrm{Var}(X)$.

6. Interpret the values of $a$, $b$, and $c$ in the context of this problem.

7. Explain briefly why a Triangular distribution is a reasonable model in this setting.

8. Compute the probability that the installation takes at most $5$ hours, that is, compute

$$
P(X \leq 5).
$$

9. Compute the probability that the installation takes between $5$ and $8$ hours, that is, compute

$$
P(5 \leq X \leq 8).
$$

10. Briefly explain how this model could be used in project planning or sensitivity analysis.


### Solutions for Q5: Expected Project Completion Time (Triangular Distribution)

**1. Support of $X$:**
The support is $[a, b] = [4, 10]$.

**2. Probability Density Function $f_X(x)$:**
With $a=4, b=10, c=6$:
$$f_X(x) = \begin{cases} \frac{2(x-4)}{(10-4)(6-4)} = \frac{x-4}{6} & 4 \le x < 6 \\ \frac{2(10-x)}{(10-4)(10-6)} = \frac{10-x}{12} & 6 \le x \le 10 \\ 0 & \text{otherwise} \end{cases}$$

**3. Verification:**
$\int_4^{10} f_X(x) dx = \int_4^6 \frac{x-4}{6} dx + \int_6^{10} \frac{10-x}{12} dx = [\frac{(x-4)^2}{12}]_4^6 + [-\frac{(10-x)^2}{24}]_6^{10} = \frac{4}{12} + \frac{16}{24} = \frac{1}{3} + \frac{2}{3} = 1$.

**6. Context of $a, b, c$:**
* $a=4$: Optimistic time (minimum possible duration).
* $b=10$: Pessimistic time (maximum possible duration).
* $c=6$: Most likely time (mode of the distribution).

**7. Reasoning for Triangular Distribution:**
It is used when data is scarce but experts can provide three estimates (best-case, worst-case, most-likely). It is simpler than the Beta/PERT distribution but captures the essential skewness of project tasks.

In [3]:
from scipy.stats import triang

a, b, c = 4, 10, 6

# Scipy triang parameters: c_scipy = (c-a)/(b-a), loc=a, scale=b-a
c_scipy = (c - a) / (b - a)
loc = a
scale = b - a

# 4. Expectation E[X] = (a + b + c) / 3
expectation_q5 = (a + b + c) / 3

# 5. Variance Var(X) = (a^2 + b^2 + c^2 - ab - ac - bc) / 18
variance_q5 = (a**2 + b**2 + c**2 - a*b - a*c - b*c) / 18

# 8. P(X <= 5)
p_le_5 = triang.cdf(5, c_scipy, loc, scale)

# 9. P(5 <= X <= 8)
p_range_q5 = triang.cdf(8, c_scipy, loc, scale) - triang.cdf(5, c_scipy, loc, scale)

print(f"4. Expectation E[X]: {expectation_q5:.4f}")
print(f"5. Variance Var(X): {variance_q5:.4f}")
print(f"8. P(X <= 5): {p_le_5:.4f}")
print(f"9. P(5 <= X <= 8): {p_range_q5:.4f}")

4. Expectation E[X]: 6.6667
5. Variance Var(X): 1.5556
8. P(X <= 5): 0.0833
9. P(5 <= X <= 8): 0.7500


**10. Practical Use:**
This model is used in **Monte Carlo simulations** for project planning to estimate the probability of finishing a project by a deadline (e.g., $P(X \le T)$). **Sensitivity analysis** can show how changing the 'most likely' estimate ($c$) affects the overall project schedule risk.

#Q6) Expected Project Completion Time Using PERT

A project manager is estimating the time required to complete a critical software integration task. Let $X$ denote the number of days required to complete the task. Based on expert judgment, the shortest possible time is $a=8$ days, the longest possible time is $b=20$ days, and the most likely completion time is $c=11$ days. Assume that $X$ follows a PERT distribution with parameters $(a,b,c)$.

The PERT distribution is defined as a Beta distribution scaled to the interval $[a,b]$, with internal shape parameters

$$
\alpha = 1 + 4\left( \frac{c-a}{b-a} \right)
$$

and

$$
\beta = 1 + 4\left( \frac{b-c}{b-a} \right).
$$

Its probability density function is

$$
f(x)=\frac{(x-a)^{\alpha-1}(b-x)^{\beta-1}}{B(\alpha,\beta)(b-a)^{\alpha+\beta-1}}, \qquad a \leq x \leq b,
$$

where

$$
B(\alpha,\beta)=\int_0^1 t^{\alpha-1}(1-t)^{\beta-1},dt.
$$

Answer the following:

1. Compute the values of the shape parameters $\alpha$ and $\beta$.

2. Specify the support of $X$.

3. Write down the probability density function $f_X(x)$ for this model.

4. State the probability space $(\Omega,\mathfrak F,P)$ suitable for this continuous model.

5. Define $X$ as a random variable on this space.

6. Show that the expectation $E[X]$ is given by

$$
E[X]=\frac{a+4c+b}{6}.
$$

7. Compute the variance $\mathrm{Var}(X)$ using the approximation

$$
\mathrm{Var}(X)=\frac{(E[X]-a)(b-E[X])}{7}.
$$

8. Compute the approximate standard deviation $\sigma$.

9. Briefly explain why a PERT distribution may be more appropriate than a Triangular distribution in this setting byt comparing the two results.

10. Interpret the practical meaning of the values $a$, $b$, $c$, $E[X]$, and $\sigma$ in the context of project planning.


### Solutions for Q6: Software Integration (PERT Distribution)

**1. Shape Parameters $\alpha$ and $\beta$:**
Given $a=8, b=20, c=11$:
$$\alpha = 1 + 4\left( \frac{11-8}{20-8} \right) = 1 + 4\left(\frac{3}{12}\right) = 2$$
$$\beta = 1 + 4\left( \frac{20-11}{20-8} \right) = 1 + 4\left(\frac{9}{12}\right) = 4$$

**2. Support of $X$:**
The support is $[a, b] = [8, 20]$.

**3. Probability Density Function $f_X(x)$:**
$$f(x) = \frac{(x-8)^{2-1}(20-x)^{4-1}}{B(2,4)(20-8)^{2+4-1}} = \frac{(x-8)(20-x)^3}{B(2,4) \cdot 12^5}, \quad 8 \le x \le 20$$

**4 & 5. Probability Space and Random Variable:**
* **$\Omega = [8, 20]$**, with the Borel sigma-algebra.
* **$X$** is the identity mapping $X(\omega) = \omega$.

**6. Show Expectation $E[X] = \frac{a+4c+b}{6}$:**
The expectation of a Beta distribution on $[0,1]$ is $\frac{\alpha}{\alpha+\beta}$. For a PERT distribution scaled to $[a,b]$, the mean is $a + (b-a)\frac{\alpha}{\alpha+\beta}$. Substituting $\alpha$ and $\beta$:
$$E[X] = a + (b-a)\frac{1+4\frac{c-a}{b-a}}{2+4} = a + \frac{(b-a) + 4(c-a)}{6} = \frac{6a + b - a + 4c - 4a}{6} = \frac{a + 4c + b}{6}$$

In [4]:
import numpy as np

a6, b6, c6 = 8, 20, 11

# 6. Compute Expectation
expectation_q6 = (a6 + 4*c6 + b6) / 6

# 7. Compute Variance using the approximation provided
# Var(X) = ((E[X] - a) * (b - E[X])) / 7
variance_q6 = ((expectation_q6 - a6) * (b6 - expectation_q6)) / 7

# 8. Standard Deviation
std_q6 = np.sqrt(variance_q6)

print(f"1. Alpha: {1 + 4*((c6-a6)/(b6-a6)):.1f}, Beta: {1 + 4*((b6-c6)/(b6-a6)):.1f}")
print(f"6. Expectation E[X]: {expectation_q6:.2f} days")
print(f"7. Variance Var(X): {variance_q6:.4f}")
print(f"8. Standard Deviation: {std_q6:.4f} days")

1. Alpha: 2.0, Beta: 4.0
6. Expectation E[X]: 12.00 days
7. Variance Var(X): 4.5714
8. Standard Deviation: 2.1381 days


**9. Comparison:**
In Q5 (Triangular), the distribution is linear and "peaky". The PERT distribution used here is a smoother Beta curve that puts less weight on the extreme tails ($a$ and $b$) compared to the Triangular distribution, making it generally more realistic for human-estimated tasks.

**10. Interpretation:**
* **$E[X]=12$**: This is the weighted average time. Note it is higher than the mode ($c=11$) because the long tail goes up to 20.
* **$\sigma \approx 2.14$**: Represents the standard risk or uncertainty in the software task duration.

#Q7) If the PERT Expectation is to Match Our Intuition

A project manager models the duration $X$ of a task by a PERT distribution with parameters $a$, $b$, and $c$, where $a$ is the minimum time, $b$ is the maximum time, and $c$ is the most likely time.

Suppose the minimum time is $a=5$ days and the most likely time is $c=11$ days. If the expected time computed from the PERT model is also to be the same as the most likely time:

1. Determine the value of the maximum time $b$.

2. What happens to the variance?

### Solutions for Q7: Matching Expectation to Intuition

**1. Determine the value of maximum time $b$:**
We are given $a=5$, $c=11$, and the condition $E[X] = c$.
Using the PERT expectation formula:
$$E[X] = \frac{a + 4c + b}{6}$$
Setting $E[X] = c$:
$$c = \frac{a + 4c + b}{6}$$
$$6c = a + 4c + b$$
$$2c = a + b$$
$$b = 2c - a$$
Substituting the given values:
$$b = 2(11) - 5 = 22 - 5 = 17$$
So, the maximum time must be **17 days**.

**2. What happens to the variance?**
When $E[X] = c$, the distribution becomes symmetric ($c$ is exactly halfway between $a$ and $b$).
In a symmetric PERT distribution, the shape parameters are equal ($\alpha = \beta = 3$).
The variance in this specific case is:
$$\text{Var}(X) = \frac{(c-a)(b-c)}{7} = \frac{(11-5)(17-11)}{7} = \frac{6 \cdot 6}{7} = \frac{36}{7} \approx 5.143$$